target/snitch_cluster/sw/apps/snax-streamer-gemm-conv-simd/data/datagen.py

```diff
@@ -0,0 +1,668 @@
+    out_channels, kernel_height, kernel_width, _ = kernel.shape
+    stride_h, stride_w = stride
+    pad_h, pad_w = padding
+
+    # Calculate the size of the output feature map
+    out_height = (in_height + 2 * pad_h - kernel_height) // stride_h + 1
+    out_width = (in_width + 2 * pad_w - kernel_width) // stride_w + 1
+
+    # Apply zero padding to the input data
+    input_data_padded = np.pad(
+        input_data, ((0, 0), (pad_h, pad_h), (pad_w, pad_w), (0, 0)), mode="constant"
+    )
+
+    # Initialize the im2col matrix
+    im2col_matrix = np.zeros(
+        (batch_size, out_height * out_width, in_channels * kernel_height * kernel_width)
+    )
+
+    # Perform the im2col transformation on the input data
+    for b in range(batch_size):
+        for oh in range(out_height):
+            for ow in range(out_width):
+                # Calculate the input region
+                ih_start = oh * stride_h
+                ih_end = ih_start + kernel_height
+                iw_start = ow * stride_w
+                iw_end = iw_start + kernel_width
+
+                # Slice and extract the input region
+                input_region = input_data_padded[b, ih_start:ih_end, iw_start:iw_end, :]
+
+                # Flatten the input region into a 1D vector and add it to the
+                # corresponding position in the im2col matrix
+                im2col_matrix[b, oh * out_width + ow, :] = input_region.reshape(-1)
+
+    im2col_matrix = im2col_matrix.reshape(batch_size * out_height * out_width, -1)
+    im2col_kernel = kernel.reshape(out_channels, -1).T
+
+    return im2col_matrix, im2col_kernel
+
+
+# Golden model in python
+# Golden model in python
+def block_gemm_golden_model(
+    m, k, n, row, size, col, a, b, subtraction_a, subtraction_b, c
+):
+    d = np.zeros(m * row * n * col, dtype=(np.int32))
+    for mm in range(m):
+        for nn in range(n):
+            for kk in range(k):
+                for rr in range(row):
+                    for cc in range(col):
+                        for ss in range(size):
+                            c_index = (
+                                mm * n * row * col + nn * row * col + rr * col + cc
+                            )
+                            a_index = (
+                                mm * k * row * size + kk * row * size + rr * size + ss
+                            )
+                            b_index = (
+                                nn * k * size * col + kk * size * col + cc * size + ss
+                            )
+                            d[c_index] = d[c_index] + (a[a_index] - subtraction_a) * (
+                                b[b_index] - subtraction_b
+                            )
+    d = np.add(c, d)
+    return d
+
+
+def data_reshuffler_golden_model(
+    tempLoop0,
+    tempLoop1,
+    spatial_len_0,
+    spatial_len_1,
+    tempStride0,
+    tempStride1,
+    spatialStride0,
+    spatialStride1,
+    data,
+    int32=False,
+):
```

Comment:
I think at this point, we need to make the data reshuffles or whatever useful functions in a single python script to make it easy for you 😄

Reply:
I meant next PR you do that I guess? Or do you have more to fix so far?